### ISSUE TITLE
Reading jpg image from disk and passing as input to a pretrained model loaded from disk

### ISSUE THREAD

**Issue by alimoeeny** [galeone/tfgo#78]
Hi @galeone  and friends, 

I have a pretrained model saved on disk, and a jpeg image saved on disk as well. 
I load the model and read the image and pass the image to the model. 
If I do this using python + tensorflow + Pillow I get a completely different result compared to when I use tfgo. 
Here is my go code:

``` go
        model := tg.LoadModel(modelDirectory), []string{"serve"}, nil)

        img := tfimg.Read(root, imagePathOnLocalDisk, 3)
	batch := tg.Batchify(root, []tf.Output{img.Value()})
	results := tg.Exec(root, []tf.Output{batch}, nil, &tf.SessionOptions{})
	inputImage, _ := tf.NewTensor(results[0].Value())

	output := model.Exec([]tf.Output{
		model.Op("StatefulPartitionedCall", 0),
	}, map[tf.Output]*tf.Tensor{
		model.Op("serving_default_input_2", 0): inputImage,
	})
```

Any comments, feedback, recommendations would be very much appreciated. 
- How do you verify that the model behaves the same way when it is moved from training (using python on training env with GPU and specific arch etc) to inference (using go, using cpu etc)? 
- what does that "serve" SessionOption does? I copied it from a code sample in the docs or another issue here I think? Any other options I need to pass?

**Comment by galeone** [comment]
Hi @alimoeeny 

The `image.Read*` functions automatically do these things:

- Read the file
- Decode using the correct codec (jpeg,png,gif, ...). Create a  `H,W,D` tensor
- Add the batch dimension. Create a  `1, H, W, D` tensor.
- Convert the image from dtype `uint8` to dtype `float32`
- Scales the image in the `[0,1]` range.

Are you sure your python code is doing the same preprocessing the `Read` function is doing?

Moreover I guess the line `batch := tg.Batchify(root, []tf.Output{img.Value()})` can be removed since `img` should already be a tensor with a batch dimension of 1.

For answering your questions:

1. I suggest first trying with Python code. Once you have your SavedModel, try to import it with [tf.saved_model.load](https://www.tensorflow.org/api_docs/python/tf/saved_model/load), and feed the model in the same way you'd do it in Go. You can specify with `tf.device` a context manager for the execution of the loaded model, and make it work on your CPU, to simulate what will happen by default in Go. I'd just print out the numbers (e.g the predictions or whatever the output is).
I do the same thing in Go and verify that I can get the same numbers.
2. "serve" is the default tag contained in an exported SavedModel. You can export the saved model with `saved_model_cli` (via cli) to analyze its content. A saved model can have multiple tags - by default, there's a single tag "serve" that contains your model. This tag can be used for versioning the model or for having multipel version of the same model inside the SavedModel (very rare scenario). You can find more info  [on the doc](https://www.tensorflow.org/guide/saved_model)

**Comment by alimoeeny** [comment]
Thank you very much @galeone  for you VERY FAST response. 
I think you identified the problem. 
In python I don't "Scale the image in the `[0,1]` range" before passing it to the model. (if I am not mistaken the model I am working with has a preprocessing stage that normalizes the image and scales it and all. 
- I modified the python code to normalize the image to `[0,1]` and got an answer close to what I get on the go side. 
- I modified the `tfgo` code to scale to `[0,255]`range but the answer was NOT close to the python's original output. 

I think I am going to train a simple model with scaled inputs and see if I can get the same answer on both python and tfgo side.

**Comment by galeone** [comment]
I'm glad I helped :+1: 

On a side note, I usually recommend also exporting the pre-processing steps inside the SavedModel.

This forces you to define everything using pure TensorFlow functions (e.g. `tf.image.resize` and not pillow stuff) that's the *only* way to have the very same behavior in Python and in other languages. For example, the pillow resize produces a completely different representation (the pixel values!) of the OpenCV resize, and both produce a different value from the `tf.image.resize`  method. Portability is possible only by writing pure TensorFlow programs - otherwise, you end up in a rabbit hole I don't recommend (at work we analyzed thing thing and wrote an article about it https://blog.zuru.tech/blog/the-dangers-behind-image-resizing)
